File: 2019/Round B/building_palindromes.cpp
#include <stdio.h>
#include <vector>
#include <set>
using namespace std;
int main() {
	int t, l, r, n, q, c = 1;
	vector<vector<int>>arr;
	scanf("%d", &t);
	while (t--) {
		scanf("%d %d\n", &n, &q);
		arr = vector<vector<int>>(27, vector<int>(n+1, 0));
		set<char> se;
		char rune;
		for (int i = 1; i <= n; i++) {
			scanf("%c", &rune);
			se.insert(rune);
			for (int j = 0; j < 27; j++) {
				arr[j][i] += arr[j][i - 1];
			}
			arr[rune - 'A'][i] ++;
		}
		int ans = 0;
		while (q--) {
			int abc[28] = {};
			scanf("%d %d", &l, &r);
			int isOdd = 0;
			for (auto i = se.begin(); i != se.end(); i++) {				
				isOdd += !((arr[*i - 'A'][r] - arr[*i - 'A'][l-1]) % 2)? 0 : 1;				
			}
			if (isOdd <= 1)
				ans++;
		}
		printf("Case #%d: %d\n", c++, ans);
	}
	return 0;
}

File: 2019/Round B/building_palindromes.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var reader = bufio.NewReader(os.Stdin)
var writer  = bufio.NewWriter(os.Stdout)
func printf(f string, a ...interface{}) { fmt.Fprintf(writer, f, a...) }
func scanf(f string, a ...interface{}) { fmt.Fscanf(reader, f, a...) }

func main(){
	var t, l, r, n, q int
	c := 1

	scanf("%d\n", &t)
	for ; t != 0; t-- {
		scanf("%d %d\n", &n, &q)

		arr := make([][]int, 27)
		for i := 0; i < 27; i++ {
			arr[i] = make([]int, n+1)
		}

		var token rune
		set := make(map[rune]bool)

		for i := 1; i <= n; i++ {
			scanf("%c", &token)
			set[token] = true
			for j := 0; j < 27; j++ {
				arr[j][i] += arr[j][i - 1]
			}
			arr[int(token) - int('A')][i]++
		}
		ans := 0

		scanf("\n")
		for ; q != 0; q-- {
			scanf("%d %d\n", &l, &r)
			isOdd := 0
			for i := 0; i < 27; i++ {
				if set[token] == true {
					if ((arr[i][r] - arr[i][l-1]) % 2) != 0 {
						isOdd++
					}
				}
			}

			if isOdd <= 1 {
				ans++
			}
		}
		printf("Case #%d: %d\n", c, ans)
		c++
	}
}
